lnpbp-0021.md

```diff
@@ -56,102 +56,111 @@ Interface specification is the following Contractum code:
+data EngravingData :: 
+    appliedTo TokenId, 
+    content EmbeddedMedia
 
-data Media ::
-    type MimeType,
+data EmbeddedMedia ::
+    type RGBTypes.MediaType,
     data [Byte]
 
 data Attachment ::
-    type MimeType,
+    type RGBTypes.MediaType,
     digest: [U8 ^ 32] -- this can be any type of 32-byte hash, like SHA256(d), BLACKE3 etc
 
-data POR :: -- proof of reserves
-    reserves TxOut,
-    proof [Byte] -- schema-specific proof 
-
-data Token ::
+data TokenData ::
     id TokenId,
-    name [Ascii ^ 1..40],
-    details [Unicode ^ 40..256]?,
-    preview Media?, -- always embedded preview media < 64kb
-    media Attachment?, -- external media which is the main media for the token
+    ticker RGBTypes.Ticker?,
+    name RGBTypes.Name?,
+    details RGBTypes.Details?,
+    -- always-embedded preview media < 64kb
+    preview EmbeddedMedia?,
+    -- external media which is the main media for the token
+    media Attachment?,
     attachments { U8 -> ^ ..20 Attachment } -- auxiliary attachments by type (up to 20 attachments)
-    reserves POR? -- output containing locked bitcoins; how reserves are
-                  -- proved is a matter of a specific schema implementation
-
-data Denomination :: 
-    ticker [Ascii ^ 1..8],
-    name [Ascii ^ 1..40],
-    details [Unicode ^ 40..256]?,
-    contract [Unicode]??,
+    -- output containing locked bitcoins; how reserves are proved is a matter
+    -- of a specific schema implementation
+    reserves RGBTypes.PoR? 
 
  -- each attachment type is a mapping from attachment id 
  -- (used as `Token.attachments` keys) to a short Ascii string
  -- verbally explaining the type of the attachment for the UI
  -- (like "sample" etc).
-data AttachmentType :: id U8, description [Ascii ^ 1..20]
-
+data AttachmentType :: 
+    id U8, 
+    description [Ascii ^ 1..20]
 
 interface RGB21
-    global Name :: Denomination
-    global Tokens :: Token+
-    global Engravings :: Engraving+
+    global name :: RGBTypes.AssetNaming
+    
+    -- Contract text is separated from the name since it must not be
+    -- changeable by the issuer.
+    global terms :: RGBTypes.RicardianContract
+
+    -- Data for all issued tokens
+    global tokensData+ :: TokenData
+    global engravingData+ :: EngravingData
     global isFractional :: Bool
-    global AttachmentTypes :: AttachmentType+
-
-    owned Allocations+ :: Allocation
-    owned IssueRight* :: Amount
-    owned DenominationRight?
-    -- returns information about known circulating supply
-    read supplyKnown :: Amount
-        count Self.state["Tokens"]
-        -- sum Self.ops["issue"]..closed["usingRight"].state ?? 0
-
-    -- maximum possible asset inflation
-    read supplyLimit :: Amount
-        sum Self.IssueRight..state ?? 0
-
-    read isCirculationKnown :: Bool
-        all Self.ops["issue"]..stateKnown
-
-    op transfer    :: inputs [Allocation+] -> beneficiaries [Allocation]
+    global attachmentTypes+ :: AttachmentType
```

Comment:
Why are we using `AttachmentType` as a map for attachments? Couldn't we simply add an optional description field to the `Attachment` type?

Reply:
Nope, since it holds a list of the allowed media types